runtime/vm/UpcallThunkMem.cpp

```diff
@@ -0,0 +1,274 @@
+
+	if (NULL != thunkAddress) {
+		/* Flash the generated thunk to the memory */
+		j9cpu_flush_icache(thunkAddress, data->thunkSize);
+	}
+}
+
+/**
+ * @brief Allocate a piece of thunk memory with a given size from the existing virtual memory block
+ *
+ * @param data a pointer to J9UpcallMetaData
+ * @return the start address of the upcall thunk memory, or NULL on failure.
+ */
+void *
+allocateUpcallThunkMemory(J9UpcallMetaData *data)
+{
+	J9JavaVM * vm = data->vm;
+	void *thunkAddress = NULL;
+
+	Assert_VM_true(data->thunkSize > 0);
+
+	omrthread_monitor_enter(vm->thunkHeapWrapperMutex);
+	if (NULL == vm->thunkHeapWrapper) {
+		thunkAddress = allocateThunkHeap(data);
+	} else {
+		thunkAddress = subAllocateThunkFromHeap(data);
+	}
+	omrthread_monitor_exit(vm->thunkHeapWrapperMutex);
+
+	return thunkAddress;
+}
+
+/**
+ * Suballocate a piece of thunk memory on the heap and return the
+ * allocated address on success; otherwise return NULL
+ */
+static void *
+subAllocateThunkFromHeap(J9UpcallMetaData *data)
+{
+	J9JavaVM * vm = data->vm;
+	UDATA thunkSize = data->thunkSize;
+	J9UpcallThunkHeapWrapper *thunkHeapWrapper = vm->thunkHeapWrapper;
+	J9Heap *thunkHeap = thunkHeapWrapper->heap;
+	void *subAllocThunkPtr = NULL;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	Trc_VM_subAllocateThunkFromHeap_Entry(thunkSize);
+
+	if (NULL != thunkHeap) {
+#if defined(OSX) && defined(AARCH64)
+		pthread_jit_write_protect_np(0);
+#endif
+		subAllocThunkPtr = j9heap_allocate(thunkHeap, thunkSize);
+		if (NULL != subAllocThunkPtr) {
+			J9UpcallMetaDataEntry metaDataEntry = {0};
+			metaDataEntry.thunkAddrValue = (UDATA)(uintptr_t)subAllocThunkPtr;
+			metaDataEntry.upcallMetaData = data;
+
+			/* Store the address of the generated thunk plus the corresponding metadata as an entry to
+			 * a hashtable which will be used to release the memory automatically via freeUpcallStub
+			 * in OpenJDK when their native scope is terminated or VM exits.
+			 */
+			if (NULL == hashTableAdd(thunkHeapWrapper->metaDataHashTable, &metaDataEntry)) {
+				j9heap_free(thunkHeap, subAllocThunkPtr);
+				subAllocThunkPtr = NULL;
+				Trc_VM_subAllocateThunkFromHeap_suballoc_thunk_add_hashtable_failed(thunkSize, thunkHeap);
+#if defined(OSX) && defined(AARCH64)
+				pthread_jit_write_protect_np(1);
+#endif
+				goto done;
+			}
+			Trc_VM_subAllocateThunkFromHeap_suballoc_thunk_success(subAllocThunkPtr, thunkSize, thunkHeap);
+		} else {
+			Trc_VM_subAllocateThunkFromHeap_suballoc_thunk_failed(thunkSize, thunkHeap);
+		}
+#if defined(OSX) && defined(AARCH64)
+		pthread_jit_write_protect_np(1);
+#endif
+	}
+
+done:
+	Trc_VM_subAllocateThunkFromHeap_Exit(subAllocThunkPtr);
+	return subAllocThunkPtr;
+}
+
+/**
+ * The memory will be allocated using vmem, and an attempt will be made to
+ * suballocate the thunk within that memory. If the overhead of omrheap
+ * precludes using suballocation, omrheap will not be used and the memory
+ * will be used directly instead.
```

Comment:
it looks like this function is always suballocating from the heap. If would it be simper to use `subAllocateThunkFromHeap` for sub allocation and `allocateThunkHeap` can just create the thunkHeap?

Reply:
Technically it should be feasible this way to avoid duplicate code in these two functions but I'd like to double-check the code to see whether it can be modified to work in the same way as it is supposed to be.

Reply:
I've modified the code (and verified on AIX & pLinux) in which `subAllocateThunkFromHeap` invokes `allocateThunkHeap` to create the thunk heap at the beginning and allocates the thunk from the heap afterwards so as to avoid duplicate code in there.

Reply:
@knn-k, please double-check the code in `UpcallThunkMem.cpp` to see whether `pthread_jit_write_protect_np()` is at the right place for Aarch64 given the code around there has been modified.

Reply:
They look OK to me, although I haven't tried the updated code on AArch64 macOS.

Background:
You need to call `pthread_jit_write_protect_np()` before and after you write to a memory block that is allocated with `J9PORT_VMEM_MEMORY_MODE_EXECUTE`.  `j9heap_*` functions writes to such a memory block allocated for the thunk memory, and that requires `pthread_jit_write_protect_np()` around the calls to `j9heap_*`.

https://github.com/eclipse/omr/pull/6670 also needs to be merged for AArch64 macOS.